src/core/coap/coap.cpp

```diff
@@ -456,15 +628,465 @@ void CoapBase::DequeueMessage(Message &aMessage)
+        FreeLastBlockResponse();
+        error = OT_ERROR_NONE;
+    }
+
+exit:
+    if (error != OT_ERROR_NONE && error != OT_ERROR_BUSY && response != NULL)
+    {
+        response->Free();
+    }
+
+    return error;
+}
+
+otError CoapBase::ProcessBlock2Request(Message &               aMessage,
+                                       const Ip6::MessageInfo &aMessageInfo,
+                                       const Resource &        aResource)
+{
+    otError        error                                        = OT_ERROR_NONE;
+    Message *      response                                     = NULL;
+    uint8_t        buf[OPENTHREAD_CONFIG_COAP_MAX_BLOCK_LENGTH] = {0};
+    uint16_t       bufLen                                       = 0;
+    bool           moreBlocks                                   = false;
+    uint8_t *      optionBuf                                    = NULL;
+    OptionIterator iterator;
+
+    SuccessOrExit(error = aMessage.ReadBlockOptionValues(OT_COAP_OPTION_BLOCK2));
+
+    otLogInfoCoap("Request for Block2 Nr. %d, Size: %d bytes received", aMessage.GetBlockWiseBlockNumber(),
+                  otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize()));
+
+    if (aMessage.GetBlockWiseBlockNumber() == 0)
+    {
+        aResource.HandleRequest(aMessage, aMessageInfo);
+        ExitNow();
+    }
+
+    // Set up next response
+    VerifyOrExit((response = NewMessage()) != NULL, error = OT_ERROR_NO_BUFS);
+    response->Init(OT_COAP_TYPE_ACKNOWLEDGMENT, OT_COAP_CODE_CONTENT);
+    response->SetMessageId(aMessage.GetMessageId());
+
+    bufLen = otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize());
+    SuccessOrExit(error = aResource.mTransmitHook(aResource.mContext, buf,
+                                                  otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize()) *
+                                                      aMessage.GetBlockWiseBlockNumber(),
+                                                  &bufLen, &moreBlocks));
+
+    response->SetMoreBlocksFlag(moreBlocks);
+    if (moreBlocks)
+    {
+        switch (bufLen)
+        {
+        case 1024:
+            response->SetBlockWiseBlockSize(OT_COAP_OPTION_BLOCK_SZX_1024);
+            break;
+        case 512:
+            response->SetBlockWiseBlockSize(OT_COAP_OPTION_BLOCK_SZX_512);
+            break;
+        case 128:
+            response->SetBlockWiseBlockSize(OT_COAP_OPTION_BLOCK_SZX_128);
+            break;
+        case 64:
+            response->SetBlockWiseBlockSize(OT_COAP_OPTION_BLOCK_SZX_64);
+            break;
+        case 32:
+            response->SetBlockWiseBlockSize(OT_COAP_OPTION_BLOCK_SZX_32);
+            break;
+        case 16:
+            response->SetBlockWiseBlockSize(OT_COAP_OPTION_BLOCK_SZX_16);
+            break;
+        default:
+            error = OT_ERROR_INVALID_ARGS;
+            ExitNow();
+            break;
+        }
```

Comment:
Just thinking about this block… if I understand what this is doing correctly, this is allowing the transmit hook to change the block size.  I'm not sure if this is only intended to be used on the very first block, it does look like this would allow the hook to change the block size mid-flight, which could be a risky thing to do with some CoAP implementations.

It might be worth setting this some point earlier in the request sequence just once… that or if that's not possible, we should make it very clear to developers that they should not change `bufLen` from inside their transmit hooks after the first block has been sent (and thus block-size has been negotiated).

Reply:
I've added a warning to the description of the receive hook. Otherwise i would have to change the block size with no means of informing the user.